src/opaque.rs

```diff
@@ -104,37 +112,39 @@ where
     type Error = ProtocolError;
 
     fn try_from(second_message_bytes: &[u8]) -> Result<Self, Self::Error> {
-        let checked_slice = check_slice_size(
-            second_message_bytes,
-            Grp::ElemLen::to_usize(),
-            "second_message_bytes",
-        )?;
-        // Check that the message is actually containing an element of the
-        // correct subgroup
-        let arr = GenericArray::from_slice(&checked_slice);
-        let beta = Grp::from_element_slice(arr)?;
-        Ok(Self { beta })
+        Self::deserialize(second_message_bytes)
     }
 }
 
 impl<Grp> RegisterSecondMessage<Grp>
 where
     Grp: Group,
 {
-    /// byte representation for the registration response message
-    fn to_bytes(&self) -> Vec<u8> {
-        self.beta.to_arr().to_vec()
-    }
-
     /// Serialization into bytes
     pub fn serialize(&self) -> Vec<u8> {
         let mut registration_response: Vec<u8> = Vec::new();
-        registration_response.extend_from_slice(&serialize((&self.to_bytes()).to_vec(), 2));
-        registration_response.extend_from_slice(&serialize(Vec::new(), 2));
+        registration_response.extend_from_slice(&serialize(self.beta.to_arr().to_vec(), 2));
+        registration_response.extend_from_slice(&serialize(self.server_s_pk.clone(), 2));
 
-        // TODO: The following should not be hardcoded, but instead be customizable
-        registration_response.extend_from_slice(&[1u8, CredentialType::SkU as u8 + 1]);
-        registration_response.extend_from_slice(&[1u8, CredentialType::PkS as u8 + 1]);
+        // Handle ecf serialization
+        let secret_credentials: Vec<u8> = self
+            .ecf
+            .secret_credentials
+            .iter()
+            .map(|&x| x as u8 + 1)
+            .collect();
+        let cleartext_credentials: Vec<u8> = self
+            .ecf
+            .cleartext_credentials
+            .iter()
+            .map(|&x| x as u8 + 1)
+            .collect();
+        let ecf_serialized = [
+            serialize(secret_credentials, 1),
```

Comment:
Two length-prefixed lists of credential types, followed by the payload are worrisome, in that they don't follow the TLS convention of `[(length-prefix, payload), ... (length-prefix, payload), ]`. Now I'm ok with that in so far as we're introducing a new and fancy type, but as long as we're going there ... might as well go with a bitmap.

Reply:
Can you clarify? I am not sure I understand, I believe this is following the spec as dictated here: https://github.com/cfrg/draft-irtf-cfrg-opaque/blame/master/draft-irtf-cfrg-opaque.md#L569-L570

What is the deviation?

Reply:
I was confused, what you're dealing with here is not a list, but a single credential formats declaration. All is right with the universe.